django/db/backends/postgresql/base.py

```diff
@@ -228,7 +210,12 @@ def pool(self):
             pool = ConnectionPool(
                 kwargs=connect_kwargs,
                 open=False,  # Do not open the pool during startup.
-                configure=self._configure_connection,
+                configure=lambda conn: self._configure_connection(
+                    conn,
+                    self.ops,
```

Comment:
`self.ops` still was constructed with `self` meaning the pool keeps a strong reference to the `DatabaseWrapper` that was calling it. Can we think (just think, don't code -- not that you have to rework it a 100 times) of any other option to make this less "bad".

One idea would be to construct a new `DatabaseWrapper` instance just for this, ie:
```python
database_wrapper = self.__class__(self.settings_dict, self.alias)
```
and then pass `database_wrapper.ops` into `_configure_connection` (only for the pool). This would have the benefit that the chances that this wrapper ever has an open connection to the db are minimal and would reduce potential side effects.

Is it nice? No it is ugly as hell but I am not exactly sure which other options we have. Open for creative ideas!

Reply:
Another option is also that we say all this ugliness is not worth it and we go back to normal methods that access `self` as wanted and we fix it if it ever really becomes a problem.

Reply:
Thank you @apollo13 for the comments. Sarah is on holiday until next week, so I’m taking over in preparation for the next release, scheduled for September 3rd. On that date, we aim to release a fix for the release blocker reported in ticket-35688.

With that in mind, IMHO, the best course of action is to have two separated but related goals regarding that report:

* Provide a minimal fix for the release blocker, ensuring that `ensure_timezone` can be annulled or overridden when inheriting from `DatabaseWrapper`.
* Re-evaluate the design of the pool implementation and explore alternatives. We can brainstorm what the ideal solution might look like, perhaps creating a new ticket for me to triage?

If I understand correctly, your point is that there might be a bootstrap configuration issue with the pool if it relies on connection information from a specific connection. This connection might have settings or configurations altered compared to the base connection settings, potentially resulting in a connection from the pool that differs in appearance or behavior from others. Is that a fair summary?

From an initial inspection of the code, I see that the instance variable `self.ops` is not strictly necessary for `ensure_timezone` and `ensure_role`, since `self.ops` is used to issue two SQL statements that could be constructed directly from `DatabaseOperations` methods that could, in turn, be made class methods as follows (hand-wavy code):
```diff
diff --git a/django/db/backends/postgresql/base.py b/django/db/backends/postgresql/base.py
index e97ab6aa89..23ddae3fd3 100644
--- a/django/db/backends/postgresql/base.py
+++ b/django/db/backends/postgresql/base.py
@@ -86,19 +86,19 @@ def _get_varchar_column(data):
     return "varchar(%(max_length)s)" % data
 
 
-def ensure_timezone(connection, ops, timezone_name):
+def ensure_timezone(connection, op_class, timezone_name):
     conn_timezone_name = connection.info.parameter_status("TimeZone")
     if timezone_name and conn_timezone_name != timezone_name:
         with connection.cursor() as cursor:
-            cursor.execute(ops.set_time_zone_sql(), [timezone_name])
+            cursor.execute(op_class.set_time_zone_sql(), [timezone_name])
         return True
     return False
 
 
-def ensure_role(connection, ops, role_name):
+def ensure_role(connection, op_class, role_name):
     if role_name:
         with connection.cursor() as cursor:
-            sql = ops.compose_sql("SET ROLE %s", [role_name])
+            sql = op_class.compose_sql_from_conn("SET ROLE %s", [role_name], connection)
             cursor.execute(sql)
         return True
     return False
@@ -260,7 +260,8 @@ class DatabaseWrapper(BaseDatabaseWrapper):
                 "settings.DATABASES is improperly configured. "
                 "Please supply the NAME or OPTIONS['service'] value."
             )
-        if len(settings_dict["NAME"] or "") > self.ops.max_name_length():
+        max_name_length = self.ops_class.max_name_length()
+        if len(settings_dict["NAME"] or "") > max_name_length:
             raise ImproperlyConfigured(
                 "The database name '%s' (%d characters) is longer than "
                 "PostgreSQL's limit of %d characters. Supply a shorter NAME "
@@ -268,7 +269,7 @@ class DatabaseWrapper(BaseDatabaseWrapper):
                 % (
                     settings_dict["NAME"],
                     len(settings_dict["NAME"]),
-                    self.ops.max_name_length(),
+                    max_name_length,
                 )
             )
         if settings_dict["NAME"]:
@@ -364,7 +365,7 @@ class DatabaseWrapper(BaseDatabaseWrapper):
         self.close_pool()
         if self.connection is None:
             return False
-        return ensure_timezone(self.connection, self.ops, self.timezone_name)
+        return ensure_timezone(self.connection, self.ops_class, self.timezone_name)
 
     def _configure_connection(self, connection):
         # This function is called from init_connection_state and from the
@@ -373,12 +374,12 @@ class DatabaseWrapper(BaseDatabaseWrapper):
         # variables.
 
         # Commit after setting the time zone.
-        commit_tz = ensure_timezone(connection, self.ops, self.timezone_name)
+        commit_tz = ensure_timezone(connection, self.ops_class, self.timezone_name)
         # Set the role on the connection. This is useful if the credential used
         # to login is not the same as the role that owns database resources. As
         # can be the case when using temporary or ephemeral credentials.
         role_name = self.settings_dict["OPTIONS"].get("assume_role")
-        commit_role = ensure_role(connection, self.ops, role_name)
+        commit_role = ensure_role(connection, self.ops_class, role_name)
 
         return commit_role or commit_tz
 
diff --git a/django/db/backends/postgresql/operations.py b/django/db/backends/postgresql/operations.py
index 4b179ca83f..a011f60a40 100644
--- a/django/db/backends/postgresql/operations.py
+++ b/django/db/backends/postgresql/operations.py
@@ -189,10 +189,15 @@ class DatabaseOperations(BaseDatabaseOperations):
             return name  # Quoting once is enough.
         return '"%s"' % name
 
+    @classmethod
+    def compose_sql_from_conn(cls, sql, params, connection):
+        return mogrify(sql, params, connection)
+
     def compose_sql(self, sql, params):
-        return mogrify(sql, params, self.connection)
+        return self.compose_sql_from_conn(sql, params, self.connection)
 
-    def set_time_zone_sql(self):
+    @classmethod
+    def set_time_zone_sql(cls):
         return "SELECT set_config('TimeZone', %s, false)"
 
     def sql_flush(self, style, tables, *, reset_sequences=False, allow_cascade=False):
@@ -273,7 +278,8 @@ class DatabaseOperations(BaseDatabaseOperations):
     def prep_for_iexact_query(self, x):
         return x
 
-    def max_name_length(self):
+    @classmethod
+    def max_name_length(cls):
         """
         Return the maximum length of an identifier.
```
What do you think? :ear: :eyes:

Reply:
> Provide a minimal fix for the release blocker, ensuring that ensure_timezone can be annulled or overridden when inheriting from DatabaseWrapper.

Yes

> Re-evaluate the design of the pool implementation and explore alternatives. We can brainstorm what the ideal solution might look like, perhaps creating a new ticket for me to triage?

I think we will leave it mostly as is (see below) unless we determine this really turns out to be a problem.

> If I understand correctly, your point is that there might be a bootstrap configuration issue with the pool if it relies on connection information from a specific connection.

It is less that it relies on the connection information from a specific connection (which in theory and practice has to be "static" anyways for the pool to work. Ie the pool won't work if you change the settings_dict passed to `BaseDatabaseWrapper` every request), but that it *keeps* a reference to a connection via the check method (unless I took a wrong turn somewhere) preventing that connection from ever getting garbage collected unless the pool object itself is closed and removed from the registry. That said this most likely will not result in a problem unless you go out of your way to do non-standard stuff. 

> What do you think? 👂 👀

Honestly to me that looks ugly beyond repair (pretty much on the same level of ugly as my own suggestions) -- if we go down that road inconsistencies will pop up everywhere :) I'd rather -- assuming of taking the path that we will fix the reference problem only if it actually becomes a problem -- make the current free functions normal member functions again and just use `self` as needed. I know it is not ideal, but I also cannot come up with code actually causing havoc in any practical case.

Reply:
@apollo13 Fair enough, thank you for your detailed answer. I agree that we can cross the "using self" bridge when we come to it.

Following your advice, I'll push a new revision with:
1. revert `_configure_connection` to its original signature: `def _configure_connection(self, connection)`
2. restore `ensure_role` to is definition before the connection pool PR landed: `def ensure_role(self):`
3. keep the new `init_connection_timezone` but make it an instance method, and call this from existing `ensure_timezone` instance method
4. use `init_connection_timezone` in `_configure_connection` (which was restored to be an instance method in step 1)

Then, the reporter of the release blocker can override `init_connection_timezone`, solving their explained need for their QuestDB wrapper.